### ISSUE TITLE
False positive report for unaligned reference in a packed struct

### ISSUE THREAD

**Issue by mripard** [rust-lang/rust#87368]
It looks like the "reference to packed field is unaligned" is reported even when a reference points to an aligned field.

Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=65b76e22bb1337278a3a64d7008a2d7e

```rust
#[repr(C, packed)]
struct FooBazUnaligned {
    bar: u8,
    baz: usize
}

#[repr(C, packed)]
struct FooBazAligned {
    bar: u64,
    baz: usize,
}

#[repr(C, packed)]
struct FooBazStillAligned {
    bar: u128,
    baz: usize,
}

fn main() {
    let foo = FooBazUnaligned { bar: 1, baz: 2 };
    let _ = &foo.baz;

    let foo = FooBazAligned { bar: 1, baz: 2 };
    let _ = &foo.baz;

    let foo = FooBazStillAligned { bar: 1, baz: 2 };
    let _ = &foo.baz;
}

```

```
warning: reference to packed field is unaligned
  --> src/main.rs:21:13
   |
21 |     let _ = &foo.baz;
   |             ^^^^^^^^
   |
   = note: `#[warn(unaligned_references)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #82523 <https://github.com/rust-lang/rust/issues/82523>
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)

warning: reference to packed field is unaligned
  --> src/main.rs:24:13
   |
24 |     let _ = &foo.baz;
   |             ^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #82523 <https://github.com/rust-lang/rust/issues/82523>
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)

warning: reference to packed field is unaligned
  --> src/main.rs:27:13
   |
27 |     let _ = &foo.baz;
   |             ^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #82523 <https://github.com/rust-lang/rust/issues/82523>
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)

warning: 3 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 1.61s
     Running `target/debug/playground`
```

However, while the baz field in the first structure is indeed unaligned, the ones in the second and third structures aren't and therefore the warning seems like a false positive?

**Comment by ChrisDenton** [comment]
The problem is that `packed` sets the alignment to 1. So this indeed may be unaligned:
```rust
#[repr(C, packed)]
struct FooBazUnAligned {
    bar: u64,
    baz: usize,
}
```

To fix this you can use `packed(n)` like so:

```rust
#[repr(C, packed(8))]
struct FooBazAligned {
    bar: u64,
    baz: usize,
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7f70eb4aeaf3d0a5e0455994e7a02d6b).

I think the warning could be improved to mention this possibility.

**Comment by mripard** [comment]
Thanks for the answer. The issue is that we might not be in total control over that structure's definition.

I'm seeing this warning on a bindgen structure, and as far as I know there's no way to control what's being set into the packed argument? Plus, that would depend on the architecture, but I guess cfg(target_pointer_width) can help

**Comment by ChrisDenton** [comment]
Ah I see what you mean. It seems the warning is still displayed even if you wrap the struct so it's properly aligned:

```rust
#[repr(C, packed)]
struct FooBazUnAligned {
    bar: u64,
    baz: usize,
}

#[repr(C, align(8))]
struct FooBazAligned {
    inner: FooBazUnAligned
}

fn main() {
    let aligned = FooBazAligned { inner: FooBazUnAligned { bar: 0, baz: 0 } };
    let _ = &aligned.inner.bar;
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3ad8de9c616cc822309cdc0edc7eba1d).

**Comment by mripard** [comment]
Yeah, my real-world issue is a bit more convoluted, but looks like [this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=26b18d06e9f5ab2d6005d5d9e7d04fb4) and a is generated through bindgen (from system headers, so not a lot we can do to change their definition) 

It seems to me that everything is aligned properly and has no chance of being unaligned, so that warning feels weird?

**Comment by SkiFire13** [comment]
> It seems to me that everything is aligned properly and has no chance of being unaligned

Why do you think so? The alignment of `Outer` is still 1

**Comment by mripard** [comment]
Outer address will still be self-aligned though, as packed only affects the padding? So the structure address should be aligned already.

And then, even though the structure fields alignment will indeed be 1, the actual offset of outer.type_ will be 0, and outer.inter will be 8 bytes, so there's no way to have a unaligned reference?

**Comment by SkiFire13** [comment]
> Outer address will still be self-aligned though, as packed only affects the padding?

No, as explained above by @ChrisDenton `packed` also sets the alignment to 1 regardless of your struct's fields. 

> and outer.inter will be 8 bytes, so there's no way to have a unaligned reference

That's an offset from a 1 byte aligned position, your alignment is still 1 in the end. 

For example if the address of an `Outer` was 7, which is 1 byte aligned, as any address, then the address of `inter` would be 15, which is clearly not 8 bytes aligned.

**Comment by mripard** [comment]
> > and outer.inter will be 8 bytes, so there's no way to have a unaligned reference
> 
> That's an offset from a 1 byte aligned position, your alignment is still 1 in the end.
> 
> For example if the address of an `Outer` was 7, which is 1 byte aligned, as any address, then the address of `inter` would be 15, which is clearly not 8 bytes aligned.

My point is that if packed is anything like C, there's no way to get that address alignment of 7 in the first place and thus everything else falls into place nicely. Also, even a "standard" Rust struct allocated at an unaligned address would result in unaligned references. 

The documentation only mentions that it modifies the padding of the structure, and not the address alignment: https://doc.rust-lang.org/nomicon/other-reprs.html#reprpacked

And trying to test all those expectations seems to work fine as well: [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c32b97eb52d24025b6f7f825d4209050)

**Comment by SkiFire13** [comment]
> My point is that if packed is anything like C, there's no way to get that address alignment of 7 in the first place

`__attribute__ ((__packed__))` is C also results in an alignment of 1 though ([godbolt example](https://godbolt.org/z/3aK5PGsdE))

> Also, even a "standard" Rust struct allocated at an unaligned address would result in unaligned references.

But 7 **is an aligned address** for `Outer`, that's the problem.

> The documentation only mentions that it modifies the padding of the structure, and not the address alignment: https://doc.rust-lang.org/nomicon/other-reprs.html#reprpacked

In the first line it clearly says "and only align the type to a byte"

> And trying to test all those expectations seems to work fine as well: Playground

That's only because the stack on x86 happens to be aligned to 8 bytes anyway. You can clearly see an instance of `Outer` being only 1 byte aligned if you placed it in a struct like `struct Oops { _u8: u8, outer: Outer }`: [Playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=87c48d9e01dbdd029dbc401671dd3874)

**Comment by mripard** [comment]
> > My point is that if packed is anything like C, there's no way to get that address alignment of 7 in the first place
> 
> `__attribute__ ((__packed__))` is C also results in an alignment of 1 though ([godbolt example](https://godbolt.org/z/3aK5PGsdE))

Yeah, but it's the type alignment itself, not its address?

> > Also, even a "standard" Rust struct allocated at an unaligned address would result in unaligned references.
> 
> But 7 **is an aligned address** for `Outer`, that's the problem.
> 
> > The documentation only mentions that it modifies the padding of the structure, and not the address alignment: https://doc.rust-lang.org/nomicon/other-reprs.html#reprpacked
> 
> In the first line it clearly says "and only align the type to a byte"
> 
> > And trying to test all those expectations seems to work fine as well: Playground
> 
> That's only because the stack on x86 happens to be aligned to 8 bytes anyway.

Yeah, but virtually all allocators are, on any architecture? If it wasn't, then any allocation could result on an unaligned access even if the structure was padded.

> You can clearly see an instance of `Outer` being only 1 byte aligned if you placed it in a struct like `struct Oops { _u8: u8, outer: Outer }`: [Playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=87c48d9e01dbdd029dbc401671dd3874)

That's a completely different example though? Oops._u8 is properly aligned, but Oops.outer is indeed unaligned because of packed. That never occurs in the other structures. So the warning in your last example is indeed completely right, but it doesn't say anything about my examples where that unalignment never occurs.

**Comment by SkiFire13** [comment]
> Yeah, but it's the type alignment itself, not its address?

The address just has to be a multiple of the alignment. With alignment 1 and address of 7 is perfectly fine.

> Yeah, but virtually all allocators are, on any architecture?

Not really, and even if that's the case your type may contain a type that requires an alignment bigger than 8 bytes, in which case the stack alignment won't be enough.

> If it wasn't, then any allocation could result on an unaligned access even if the structure was padded.

No, if the type has the right alignment it won't happen. Again, the problem here is that `Outer` doesn't require the proper alignment.

> but Oops.outer is indeed unaligned because of packed. 

That's not true. `Oops` is not marked as `packed`, so its fields must be properly aligned.

**Comment by RalfJung** [comment]
> Yeah, my real-world issue is a bit more convoluted, but looks like this and a is generated through bindgen (from system headers, so not a lot we can do to change their definition)
> 
> It seems to me that everything is aligned properly and has no chance of being unaligned, so that warning feels weird?

The warning is correct, I'm afraid. `Outer` has an alignment of 1 (since it is `packed`), so it could be allocated at any address, including 7. `Inner` also has an alignment of 1 since it is `packed`; `Intermediate` inherits the alignment from its only field so that's also 1.

So, I don't think there is a false positive here.
